### ISSUE TITLE
Make generated code for an interface behave the same regardless of number of implementations

### ISSUE THREAD

**Issue by kimroen** [apollographql/apollo-ios#1754]
## Feature request
I realize there's been some existing discussion regarding interfaces and the code it generates (such as #1706 and #720), but I don't think this has come up before. If it has, please point me in the right direction. Thanks!

I'd like it to be so that the generated code for querying an interface behaves the same if there are one or multiple types implementing the interface.

## Motivation
If you have a field that returns an interface, and that interface has multiple types implementing it, the fields on the struct is structured in a specific way, and some helper functions are generated to deal with the possibility of the struct being one of a few types.

However, if only one type implements that interface, the struct looks like it would look if it were a plain object type.

As an example:

```gql
interface CommentAuthor {
  username: String!
}

# This is the only type implementing CommentAuthor
type User implements CommentAuthor {
  id: ID!
  username: String!
}

type Query {
  commentAuthor(id: ID!): CommentAuthor
}
```

If I now make this query:

```gql
query GetCommentAuthor($id: ID!) {
  commentAuthor(id: $id) {
    username
    ... on User {
      id
    }
  }
}
```

I can access the `id` directly on the `CommentAuthor` as if it were a `User`. I don't want that!

If we add another type implementing `CommentAuthor`:

```gql
type DeletedUser implements CommentAuthor {
  username: String!
} 
```

Now, we suddenly get all the helpers we need (`asUser`, `asDeletedUser`), and the `id` is only available if you first make sure you're dealing with a `User`.

In order to make sure we are future proof, I want to be forced to make sure I'm dealing with a specific implementation of the interface before I can access its unique fields.

**Comment by designatednerd** [comment]
Can you share some of the generated code please? I think this may have been an "optimization" that isn't actually all that helpful in dealing with strongly-typed languages, or for forward compatibility.

**Comment by AnthonyMDev** [comment]
Thank you for bringing this to our attention. I'm not a fan of this "optimization" either. We are in the process of re-writing Code Generation from the ground up (and in Swift instead of TypeScript). It's going to be a little while before that is ready, but this will work as you expect in the new version when it's released!

**Comment by kimroen** [comment]
> Can you share some of the generated code please? I think this may have been an "optimization" that isn't actually all that helpful in dealing with strongly-typed languages, or for forward compatibility.

Thanks! For sure - I made an example repository with the examples from this issue:

- [Commit adding the initial schema and first query](https://github.com/kimroen/apollo-generation-interface-issue/commit/21823a63586f48be35b280558c547620777dab4a)
- [Commit adding the second implementer of the interface](https://github.com/kimroen/apollo-generation-interface-issue/commit/df615564a23758b70a4ed687de0c2230ac4678ef)

Hopefully that's helpful!

> Thank you for bringing this to our attention. I'm not a fan of this "optimization" either. We are in the process of re-writing Code Generation from the ground up (and in Swift instead of TypeScript). It's going to be a little while before that is ready, but this will work as you expect in the new version when it's released!

Thanks for letting me know! Do you know of any workarounds or do you have any other tips until then?

Mostly just curious if we can get around this being an issue for me and other developers on the projects (client and server) while we wait. I was considering creating a dummy second implementer of all interfaces on the client version of the schema, but that would have to be added programmatically after downloading the schema and I'm not sure that's worth it.

Thanks again!

**Comment by AnthonyMDev** [comment]
@kimroen I believe that you could get the behavior you are looking for by using `--mergeInFieldsFromFragmentSpreads = false` when running the code gen.

Right now the code gen is making the assumption that it can merge those fields in since only one type implements that interface. I agree that this is a bad assumption, and it will be changed for the new code gen. But the work around is to turn of fragment merging completely for the time being.

This may not be the behavior you want for other fragment usage in your schema though... so if that makes things worse, then I'm not sure if there is a great workaround for now.

At the end of the day, I still recommend wrapping our generated models in your own view models that are structured the way the rest of your codebase wants to consume them. Hopefully with the new codegen our models will be more user friendly, but we know that we will never be able to generate something that fulfills every team's desired usages adequately. So wrapping with your own view models will always be a good option for these situations.

**Comment by kimroen** [comment]
> @kimroen I believe that you could get the behavior you are looking for by using `--mergeInFieldsFromFragmentSpreads = false` when running the code gen.
> 
> Right now the code gen is making the assumption that it can merge those fields in since only one type implements that interface. I agree that this is a bad assumption, and it will be changed for the new code gen. But the work around is to turn of fragment merging completely for the time being.
> 
> This may not be the behavior you want for other fragment usage in your schema though... so if that makes things worse, then I'm not sure if there is a great workaround for now.

We are not intentionally using this option, and we are just getting started so I don't think it would break anything for us at the moment. However, I tried setting it to both `true` and `false` when having only one interface (the first commit I linked to above), and it doesn't seem to make a difference to the generated output at all. Am I not doing it right, or is there some other dependency that might be causing it to behave this way?

> At the end of the day, I still recommend wrapping our generated models in your own view models that are structured the way the rest of your codebase wants to consume them. Hopefully with the new codegen our models will be more user friendly, but we know that we will never be able to generate something that fulfills every team's desired usages adequately. So wrapping with your own view models will always be a good option for these situations.

This sounds very useful - are there any code examples of doing things this way? I've seen mentions of adding extensions to the generated structs in the Apollo docs, and I'm familiar with the concept of a view model, but I'm still curious if there's anything I can read/watch about the topic to pick up tips and tricks.

Thanks!

**Comment by ceeK** [comment]
Hi folks! Just wanted to add that we are also experiencing this issue but with unions that only have a single type rather than pure interfaces. 

Take for example the union and type:
```
union Media = Book

type Book {
  name: String!
}
```

Right now we've designed a schema that has a union with a single type, under the intention that we'll add new types in the future. 

I would expect Apollo to generate an interface that led to the callsite similar to:

```
media.asBook.name
```

However the codegen simplification actually gives media the `name` field. Type `Book` isn't exposed at all to developers. I'm not too fussed about the callsite here as we adapt these models into our own anyways, so all is fine there. 

The potential problem for us is that it adds an extra avenue for app failure. A normal union with multiple types is robust; if it accidentally receives a type it wasn't expecting it will just ignore it (although I realise that this shouldn't happen if we code our backend correctly). 

This assumption by the codegen library causes a failure because the autogenerated code doesn't consider the fact that more than one type can be returned for this field. In the above example it thinks everything is a `Book`. If a `Film` gets sent down the entire response fails. 

For example this JSON will cause the entire stack to fail when `union Media = Book` 
```json
"media": [
    {
        "__typename": "Book",
        "name": "Zero to One"
    },
    {
        "__typename": "Film",
        "name": "The Martian",
        "length": "124"
    }
]
```

However if the union was instead `union Media = Book | Film`, which resulted in the faulty assumption not occurring, then the Apollo parsing would still succeed if we accidentally sent down another type `AudioBook`:
```json
"media": [
    {
        "__typename": "Book",
        "name": "Zero to One"
    },
    {
        "__typename": "Film",
        "name": "The Martian",
        "length": "124"
    },
    {
        "__typename": "Audiobook",
        "name": "The Martian",
        "format": "mp3"
    }
]
``` 

I think we'll be fine but it does add another potential app breakage if the backend accidentally sends us a type that we don't request. Hopefully this should never happen but I'll rest easier once the new codegen tooling is release 😂

**Comment by kimroen** [comment]
Thanks for mentioning this @ceeK , I hadn't considered this case. Most of our error types are unions with just one type, so I'm surprised I haven't seen this.

**Comment by AnthonyMDev** [comment]
Thanks for that @ceeK! This is all good stuff to be aware of while working on the new version of Codegen!

@kimroen I'm not sure of any blog posts. @designatednerd might have an idea? But basically I just mean creating a view model that structures the data the way you want to consume it and initializing those view models by passing our generated models to them. Then you can do any custom mapping or validation you want to do there.

**Comment by calvincestari** [comment]
Hi 👋🏻 - this issue should be resolved in 1.0. 1.0 is currently in alpha releases with a full 1.0 release in H2 2022. Please take a look at the [latest alpha](https://github.com/apollographql/apollo-ios/releases) and let us know if this is still an issue for you.
